File: exoplayer/src/main/java/com/kjsc/exoplayer/ExoPlayerView.java
package com.kjsc.exoplayer;

import static com.google.android.exoplayer2.DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS;
import static com.google.android.exoplayer2.DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_MS;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.graphics.Color;
import android.net.Uri;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import com.google.android.exoplayer2.DefaultLoadControl;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.MediaItem;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.ui.AspectRatioFrameLayout;
import com.google.android.exoplayer2.ui.PlayerView;

import java.util.ArrayList;
import java.util.List;

public class ExoPlayerView extends PlayerView implements DefaultLifecycleObserver {
    ExoPlayer player;
    Uri uri = null;
    List<MediaItem> mediaItems = null;
    @Player.RepeatMode
    int repeatMode = Player.REPEAT_MODE_ONE;

    public ExoPlayerView(Context context) {
        this(context, null, 0);
    }

    public ExoPlayerView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ExoPlayerView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setBackgroundColor(Color.BLACK);
        setResizeMode(AspectRatioFrameLayout.RESIZE_MODE_FILL);
        if (getUseController()) {
            setFocusable(true);
            setFocusableInTouchMode(true);
        }
    }

    public void initPlayer() {
        if (player == null) {
            //修改默认的缓存时间 因为机顶盒配置过低 缓存过大会OOM
            DefaultLoadControl defaultLoadControl = new DefaultLoadControl.Builder()
                    .setBufferDurationsMs(
                            5000,
                            10000,
                            DEFAULT_BUFFER_FOR_PLAYBACK_MS,
                            DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS
                    )
                    .build();
            player = new ExoPlayer.Builder(getContext())
                    .setLoadControl(defaultLoadControl)
                    .build();
            player.setPlayWhenReady(true);
            if (listener != null) {
                player.addListener(listener);
            }
            player.setRepeatMode(repeatMode);
            setPlayer(player);
        }
    }

    public void setVideoURI(Uri uri) {
        System.out.println("setVideoURI uri=" + uri.toString());
        stop();
        initPlayer();
        this.uri = uri;
        player.setMediaItem(MediaItem.fromUri(uri));
        player.prepare();
        player.play();
    }

    public void prepare(Uri uri) {
        System.out.println("setVideoURI uri=" + uri.toString());
        stop();
        initPlayer();
        this.uri = uri;
        player.setPlayWhenReady(false);
        player.setMediaItem(MediaItem.fromUri(uri));
        player.prepare();
    }

    public void play() {
        player.play();
    }

    public void setVideoPath(String path) {
        setVideoURI(Uri.parse(path));
    }

    public void setMediaItems(List<MediaItem> mediaItems, int startIndex) {
        stop();
        initPlayer();
        this.mediaItems = mediaItems;
        if (mediaItems != null && mediaItems.size() > 0) {
            player.setMediaItems(mediaItems, startIndex, 0);
            player.prepare();
            player.play();
        }
    }


    public void setMediaItems(List<MediaItem> mediaItems) {
        setMediaItems(mediaItems, 0);
    }

    public void setIMediaItems(List<? extends IMediaItem> iMediaItems) {
        List<MediaItem> mediaItems = new ArrayList<>();
        for (IMediaItem iMediaItem : iMediaItems) {
            mediaItems.add(iMediaItem.getMediaItem());
        }
        setMediaItems(mediaItems);
    }

    public void setIMediaItems(List<? extends IMediaItem> iMediaItems, int startIndex) {
        List<MediaItem> mediaItems = new ArrayList<>();
        for (IMediaItem iMediaItem : iMediaItems) {
            mediaItems.add(iMediaItem.getMediaItem());
        }
        setMediaItems(mediaItems, startIndex);
    }

    public void setMediaItemsIndex(int index) {
        if (player != null && index >= 0 && index < mediaItems.size()) {
            player.seekToDefaultPosition(index);
        }
    }

    public int getMediaItemsIndex() {
        if (player != null) {
            return player.getCurrentWindowIndex();
        }
        return -1;
    }

    public void setRepeatMode(@Player.RepeatMode int repeatMode) {
        this.repeatMode = repeatMode;
        if (player != null) {
            player.setRepeatMode(repeatMode);
        }
    }

    @Override
    public boolean dispatchKeyEvent(KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_DOWN && getUseController()) {
//            if(isFull && event.getKeyCode()==KeyEvent.KEYCODE_BACK){
//                stopFull();
//                return true;
//            }
            switch (event.getKeyCode()) {
                case KeyEvent.KEYCODE_DPAD_CENTER:
                case KeyEvent.KEYCODE_ENTER:
                    View btnPlay = findViewById(R.id.exo_play);
                    View btnPause = findViewById(R.id.exo_pause);
                    if (btnPlay.isShown())
                        btnPlay.performClick();
                    else
                        btnPause.performClick();
                    break;
            }
        }
        return super.dispatchKeyEvent(event);
    }

    Player.Listener listener;

    public void addListener(Player.Listener listener) {
        this.listener = listener;
        if (player != null) {
            player.addListener(listener);
        }
    }

    public void stop() {
        try {
            if (player != null) {
                player.stop();
                player.release();
                player = null;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 让控件跟随生命周期
     *
     * @param lifecycle
     */
    public void setLifecycle(Lifecycle lifecycle) {
        if (lifecycle != null) {
            lifecycle.removeObserver(this);
            lifecycle.addObserver(this);
        }
    }

    int resumeIndex = 0;

    @Override
    public void onStart(@NonNull LifecycleOwner owner) {
        DefaultLifecycleObserver.super.onStart(owner);
        super.onResume();
    }

    @Override
    public void onResume(@NonNull LifecycleOwner owner) {
        DefaultLifecycleObserver.super.onResume(owner);
        super.onResume();
        System.out.println("OnLifecycleEvent ON_RESUME " + uri + " isResumePlay=" + isResumePlay);
        if (isResumePlay && player == null) {
            if (mediaItems != null) {
                setMediaItems(mediaItems, resumeIndex);
            } else if (uri != null) {
                setVideoURI(uri);
            }
            isResumePlay = false;
        }
    }

    @Override
    public void onPause(@NonNull LifecycleOwner owner) {
        DefaultLifecycleObserver.super.onPause(owner);
        isResumePlay = true;
        System.out.println("OnLifecycleEvent ON_PAUSE " + uri);
        resumeIndex = getMediaItemsIndex();
        stop();
    }

    public ExoPlayer getPlayer() {
        return player;
    }


    boolean isResumePlay = false;

    public boolean isResumePlay() {
        return isResumePlay;
    }

    public void setResumePlay(boolean resumePlay) {
        isResumePlay = resumePlay;
    }

    public List<MediaItem> getMediaItems() {
        return mediaItems;
    }

    public <T> ArrayList<T> getMediaItemsData() {
        ArrayList<T> list = new ArrayList<>();
        for (MediaItem mediaItem : mediaItems) {
            list.add((T) mediaItem.localConfiguration.tag);
        }
        return list;
    }

    public Uri getUri() {
        return uri;
    }

//    ViewGroup parent=null;
//    int index=0;
//    boolean isFull=false;
//    View oldFocusView=null;
//    public void full(){
//        parent= (ViewGroup) getParent();
//        index=parent.indexOfChild(this);
//        parent.removeView(this);
//        Activity activity = scanForActivity(getContext());
//        ViewGroup contentView = (ViewGroup) activity.getWindow().getDecorView();
//        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
//                ViewGroup.LayoutParams.MATCH_PARENT);
//        contentView.addView(this, params);
//        setUseController(true);
//        oldFocusView=contentView.findFocus();
//
//        setFocusable(true);
//        requestFocus();
//        isFull=true;
//    }
//
//    public void stopFull(){
//        isFull=false;
//        setUseController(false);
//        if(oldFocusView!=null){
//            oldFocusView.requestFocus();
//            oldFocusView=null;
//        }
//        setFocusable(false);
//
//        Activity activity = scanForActivity(getContext());
//        ViewGroup contentView = (ViewGroup) activity.getWindow().getDecorView();
//        contentView.removeView(this);
//        parent.addView(this,index);
//
//        parent=null;
//    }
}

File: exoplayer/src/main/java/com/kjsc/exoplayer/ExoPlayerView.kt
package com.kjsc.exoplayer

import android.app.Activity
import android.app.Application
import android.app.Service
import android.content.Context
import android.content.ContextWrapper
import android.graphics.Color
import android.net.Uri
import android.util.AttributeSet
import android.view.*
import android.widget.FrameLayout
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import com.google.android.exoplayer2.DefaultLoadControl
import com.google.android.exoplayer2.ExoPlayer
import com.google.android.exoplayer2.MediaItem
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.ui.AspectRatioFrameLayout
import com.google.android.exoplayer2.ui.PlayerView


class ExoPlayerView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
    FrameLayout(
        context, attrs, defStyleAttr
    ), DefaultLifecycleObserver {

    var player: ExoPlayer? = null
    var playerView: PlayerView
    var controller: CustomPlayerControlView
    var progressBar: CustomProgressBar


    var uri: Uri? = null
    private var mediaItems: List<MediaItem>? = null
    var repeatMode = Player.REPEAT_MODE_ONE
        set(value) {
            field = value
            player?.repeatMode = repeatMode
        }

    init {
        playerView = PlayerView(context, attrs, defStyleAttr)
        playerView.setPadding(0, 0, 0, 0)
        playerView.isFocusable = false
        val params = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        addView(playerView, params)
        controller = findViewById(R.id.exo_controller)
        progressBar = findViewById(R.id.exo_progress)
        setBackgroundColor(Color.BLACK)
        playerView.resizeMode = AspectRatioFrameLayout.RESIZE_MODE_FILL
        if (playerView.useController) {
            isFocusable = true
            isFocusableInTouchMode = true
        }
    }

    fun initPlayer() {
        destroy()
        //修改默认的缓存时间 因为机顶盒配置过低 缓存过大会OOM
        val defaultLoadControl = DefaultLoadControl.Builder()
            .setBufferDurationsMs(
                5000,
                10000,
                DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_MS,
                DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS
            )
            .build()
        player = ExoPlayer.Builder(context)
            .setLoadControl(defaultLoadControl)
            .build()
        player?.playWhenReady = true
        if (listener != null) {
            player?.addListener(listener!!)
        }
        player?.repeatMode = repeatMode
        playerView.setPlayer(player)
    }

    fun setVideoURI(uri: Uri, startPositionMs: Long = 0) {
        println("setVideoURI uri=$uri")
        initPlayer()
        this.uri = uri
        player?.setMediaItem(MediaItem.fromUri(uri), startPositionMs)
        player?.prepare()
        player?.play()
    }

    fun prepare(uri: Uri) {
        println("setVideoURI uri=$uri")
        initPlayer()
        this.uri = uri
        player?.playWhenReady = false
        player?.setMediaItem(MediaItem.fromUri(uri))
        player?.prepare()
    }

    fun play() {
        player?.play()
    }

    fun setVideoPath(path: String?) {
        setVideoURI(Uri.parse(path))
    }

    fun setMediaItems(mediaItems: List<MediaItem>?, startIndex: Int = 0, startPos: Long = 0) {
        initPlayer()
        this.mediaItems = mediaItems
        if (mediaItems != null && mediaItems.isNotEmpty()) {
            player?.setMediaItems(mediaItems, startIndex, startPos)
            player?.prepare()
            player?.play()
        }
    }

    fun getMediaItems(): List<MediaItem>? {
        return mediaItems
    }

    fun setIMediaItems(iMediaItems: List<IMediaItem>, startIndex: Int = 0, startPos: Long = 0) {
        val mediaItems: MutableList<MediaItem> = ArrayList()
        for (iMediaItem in iMediaItems) {
            mediaItems.add(iMediaItem.mediaItem)
        }
        setMediaItems(mediaItems, startIndex, startPos)
    }

    var mediaItemsIndex: Int
        get() = player?.currentMediaItemIndex ?: -1
        set(index) {
            if (player != null && index >= 0 && index < mediaItems!!.size) {
                player?.seekToDefaultPosition(index)
            }
        }

    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        if (event.action == KeyEvent.ACTION_DOWN && playerView.useController) {
            if (playerView.useController) {
                controller.show()
                val ret1 = controller.dispatchKeyEvent(event)
                val ret2 = progressBar.onKeyDown(event.keyCode, event)
                if (ret1 || ret2) {
                    return true
                }
            }
            if (event.keyCode == KeyEvent.KEYCODE_BACK) {
                if (playerView.parent != this) {
                    exitFullScreen()
                    return true
                }
            }
        }

        return super.dispatchKeyEvent(event)
    }

    fun isFullScreen(): Boolean {
        return playerView.parent != this
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {

        return super.onKeyDown(keyCode, event)
    }

    var listener: Player.Listener? = null
    fun addListener(listener: Player.Listener?) {
        this.listener = listener
        if (player != null) {
            player?.addListener(listener!!)
        }
    }

    fun destroy() {
        try {
            player?.stop()
            player?.release()
            player = null
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 让控件跟随生命周期
     *
     * @param lifecycle
     */
    fun setLifecycle(lifecycle: Lifecycle?) {
        if (lifecycle != null) {
            lifecycle.removeObserver(this)
            lifecycle.addObserver(this)
        }
    }

    var resumeIndex = 0 //记录上次播放的视频下标
    var resumePosition = 0L //记录上次播放的视频位置
    var isResumePlay = false //是否需要恢复播放
    override fun onPause(owner: LifecycleOwner) {
        playerView.onPause()
        player?.let {
            isResumePlay = true
            resumeIndex = mediaItemsIndex
            resumePosition = it.currentPosition
        }
        destroy()
    }

    override fun onResume(owner: LifecycleOwner) {
        playerView.onResume()
        println("OnLifecycleEvent ON_RESUME $uri isResumePlay=$isResumePlay")
        if (isResumePlay && player == null) {
            if (mediaItems != null) {
                setMediaItems(mediaItems, resumeIndex, resumePosition)
            } else if (uri != null) {
                setVideoURI(uri!!, resumePosition)
            }
            isResumePlay = false
        }
    }

    fun <T> getMediaItemsData(): ArrayList<T?> {
        val list = ArrayList<T?>()
        for (mediaItem in mediaItems!!) {
            list.add(mediaItem.localConfiguration!!.tag as T?)
        }
        return list
    }

    fun fullScreen() {
        val activity = getActivityFromContext(context)
        val viewGroup = activity?.window?.decorView as? ViewGroup
        if (viewGroup != null) {
            removeView(playerView)
            val params =
                ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
            viewGroup.addView(playerView, params)
            playerView.useController = true
        }
    }

    fun exitFullScreen() {
        val activity = getActivityFromContext(context)
        val viewGroup = activity?.window?.decorView as? ViewGroup
        if (viewGroup != null) {
            viewGroup.removeView(playerView)
            val params =
                ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
            addView(playerView, params)
            playerView.useController = false
        }
    }

    fun getActivityFromContext(context: Context?): Activity? {
        if (context == null) {
            return null
        }
        if (context is Activity) {
            return context
        }
        if (context is Application || context is Service) {
            return null
        }
        var c = context
        while (c != null) {
            if (c is ContextWrapper) {
                c = c.baseContext
                if (c is Activity) {
                    return c
                }
            } else {
                return null
            }
        }
        return null
    }

}